### PR TITLE
Clean up files, exports, dependencies

### PR DESCRIPTION
**Pull request by with-heart**
This PR installs and configures [`knip`](https://knip.dev/) which I then used to clean up a number of things in the repo:

- exports
  - deleted unused types/vars that were only marked as used due to `export` keyword but aren't imported/re-exported
  - removed `export` from vars/types that aren't imported/re-exported
- removed unused files
- dependencies
  - removed unused dependencies
  - removed dependencies from packages that are specified in root
  - added installed but unspecified dependencies ([phantom dependencies](https://rushjs.io/pages/advanced/phantom_deps/#phantom-dependencies)) in the appropriate places

I think it'd be cool to eventually integrate `knip` into our pipeline to be able to automate catching these things but for now doing going through them manually is a good start.

P.S.: Since stuff is changing all over the place, it might be easier to [read the changes commit-by-commit](https://github.com/statelyai/xstate/pull/4974/commits)

P.P.S.: Press <kbd>n</kbd> to go to the next commit and press <kbd>p</kbd> to go to the previous commit

### CODE REVIEW

File: packages/core/src/actors/observable.ts
Comment (davidkpiano) [inline packages/core/src/actors/observable.ts:38]: Hmm, may want to make this one exported in `index.ts`

File: packages/core/src/actors/transition.ts
Comment (davidkpiano) [inline packages/core/src/actors/transition.ts:28]: Same here; we may want this exported in `index.ts`?

File: packages/core/src/actors/observable.ts
Comment (with-heart) [reply packages/core/src/actors/observable.ts:38]: image

Caaaaaan dooooooo

I had actually gone the opposite direction and started adding all of them to `index.ts` but realized it'd be less effort to just remove them and add back the ones that are worth it.

Comment (with-heart) [reply packages/core/src/actors/observable.ts:38]: Resolved in #4976

File: packages/core/src/actors/transition.ts
Comment (with-heart) [reply packages/core/src/actors/transition.ts:28]: Resolved in #4976

File: package.json
Comment (davidkpiano) [inline package.json:33]: cc. @Andarist

Comment (with-heart) [reply package.json:33]: There wasn't actually a `scripts/dev.js` file so I was thinking the file was removed at some point but the script was forgotten

Comment (davidkpiano) [reply package.json:33]: Ah okay!

Comment (Andarist) [reply package.json:33]: ye, I used this to develop v4

**Review by davidkpiano — APPROVED** [review]
cc. @Andarist for final approval

File: .knip.jsonc
Comment (Andarist) [inline .knip.jsonc:20]: why this one has been ignored? do we explicitly depend on this anywhere? that might be a mistake

File: .knip.jsonc
Comment (Andarist) [inline .knip.jsonc:1]: can this be added as a CI check?

File: scripts/typecheck-examples.js
Comment (Andarist) [inline scripts/typecheck-examples.js:1]: i think those 3 scripts were created by @davidkpiano for manual execution - I never used them but it feels like those shouldn't be removed (unless David is OK with removing them)

Comment (davidkpiano) [reply scripts/typecheck-examples.js:1]: Yes, I use these scripts

Comment (davidkpiano) [reply scripts/typecheck-examples.js:1]: @with-heart Can you add these back? 🙏

File: .knip.jsonc
Comment (with-heart) [reply .knip.jsonc:1]: Ye that's what I'm planning to do. It's still reporting a few warnings that I need to figure out how to take care of, gonna do that in a follow-up PR

File: .knip.jsonc
Comment (davidkpiano) [inline .knip.jsonc:10]: Maybe this or similar just in case we have other scripts?
```suggestion
    "scripts/{*}.js"
```

Comment (with-heart) [reply .knip.jsonc:10]: I was thinking about that but if it had been in the config to start it wouldn't have detected the unused `dev.js` script so might be better to be more specific? Not sure

File: .knip.jsonc
Comment (with-heart) [reply .knip.jsonc:21]: It's also in `resolutions`:

https://github.com/statelyai/xstate/blob/9841fab65f65d76349386b9489413f8fd134ee84/package.json#L89

**Comment by with-heart** [comment]
The `workspaces` config I used for `examples` and `templates` was incorrect (`examples` instead of `examples/*`). Actually enabling them creates a bunch of extra work like needing to install all of their dependencies so I just removed it and will revisit later maybe

**Comment by Andarist** [comment]
> @tsconfig/svelte is confusing because of packages/xstate-svelte/test/package.json - maybe just add to ignoreDependencies

I assume this might be some issue with pkg.json lookup. Such "package.json scopes" to add `type` are... weird but I've seen them used them in the wild a couple of times already. Maybe it would be worth ignoring them in Knip? This is a documented pattern in the node.js docs.

**Comment by with-heart** [comment]
@webpro Thanks so much for taking a look! Glad you were able to fix two bugs while helping us clean up the config. And also glad to learn about pkg.pr.new, that's quite a useful tool.

I was thinking about posting on twitter to ask if I knew any `knip` experts that wanted to take a look but opening an issue got us _the_ expert. The power of open source ❤️

**Comment by with-heart** [comment]
I believe this is good to go now 😁

File: package.json
Comment (with-heart) [inline package.json:73]: I subscribed to `knip` release notifications so I can switch this to an actual version once it's published

Comment (Andarist) [reply package.json:73]: I'd prefer to block this PR on that release. I had unpleasant experiences in the past with such alternative registered - mainly CodeSandbox CI just decided to forget a package that was published using that service.

**Comment by webpro** [comment]
> > @tsconfig/svelte is confusing because of packages/xstate-svelte/test/package.json - maybe just add to ignoreDependencies
> 
> I assume this might be some issue with pkg.json lookup. Such "package.json scopes" to add `type` are... weird but I've seen them used them in the wild a couple of times already. Maybe it would be worth ignoring them in Knip? This is a documented pattern in the node.js docs.

Sorry, not sure exactly what you are referring to here. Knip follows the root `package.json#workspaces` to find workspaces (extra workspaces can be configured, though). If we ignore `packages/xstate-svelte/test/package.json` in this regard, then the easiest path is to configure `test/tsconfig.json` as the path of the TypeScript config for this workspace:

```jsonc
{
  "workspaces": {
    "packages/xstate-svelte": {
      "typescript": "test/tsconfig.json",
      // shorthand for "typescript": { "config": ["test/tsconfig.json"] }
    }
  }
}
```

File: package.json
Comment (webpro) [reply package.json:73]: Absolutely.

Comment (with-heart) [reply package.json:73]: > I'd prefer to block this PR on that release. I had unpleasant experiences in the past with such alternative registered - mainly CodeSandbox CI just decided to forget a package that was published using that service.

Damn you, Codesandbox! ✊☁️

**Comment by with-heart** [comment]
Since `knip@5.27.0` was released this morning, I've replaced `pkg.pr.new` with `^5.27.0`. Running it results in no errors so we should be good to go! 🤞

**Comment by webpro** [comment]
Thanks @with-heart! ➕❤️ 

You could also use the config from [my latest comment](https://github.com/statelyai/xstate/pull/4974#issuecomment-2241596411) to get rid of of the ignored `@tsconfig/svelte` issue, but it's minor.

**Comment by Andarist** [comment]
I guess I just don't understand why this has to be declared as an extra workspace in the first place - couldn't it be picked up automatically? It's fine as a solution right now - I just wonder how Knip's magic sauce is made and if there is any room for improvements on that front

**Comment by webpro** [comment]
> I guess I just don't understand why this has to be declared as an extra workspace in the first place - couldn't it be picked up automatically?

There's no magic here, Knip just uses `package.json#workspaces` - which does not include `packages/xstate-svelte/test`.

**Comment by Andarist** [comment]
Right, because it's not a workspace. But including it in Knip's workspaces it feels like we dilute the meaning of a workspace. Multiple `tsconfig.json`s (and `package.json`s for that matter) are fair game within a single workspace. So I just wonder if they couldn't be discovered automatically? It would require less work for the user.

**Comment by webpro** [comment]
> Right, because it's not a workspace. But including it in Knip's workspaces it feels like we dilute the meaning of a workspace.

We're not including it as a workspace, it's only Knip's TypeScript plugin that pulls in `test/tsconfig.json` instead of `./tsconfig.json` so it can extract dependencies from that file. There's nothing more to it, it's not like we're defining `./test` as a TS project or a (npm/Yarn) workspace.

> Multiple `tsconfig.json`s (and `package.json`s for that matter) are fair game within a single workspace. So I just wonder if they couldn't be discovered automatically? It would require less work for the user.

Unfortunately we're simply not there yet.

**Comment by davidkpiano** [comment]
Is this one good to go on (pending merge conflicts)? @Andarist @with-heart

**Comment by with-heart** [comment]
Yeap we're good to go!

**Review by Andarist — APPROVED** [review]

**Pull request merged**
